layers/enip.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2017 Google, Inc. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file in the root of the source
+// tree.
+
+package layers
+
+import (
+	"encoding/binary"
+	"errors"
+
+	"github.com/google/gopacket"
+)
+
+const (
+	enipMinPacketLen           int = 24
+	enipMinRegSessionPacketLen int = 4
+	enipMinSendRRDataPacketLen int = 36
+
+	ENIP_TCPPort uint16 = 44818
+	ENIP_UDPPort uint16 = 2222
+)
+
+var (
+	listServices      enipCommand = 0x0004
+	listIdentity      enipCommand = 0x0063
+	listInterfaces    enipCommand = 0x0064
+	registerSession   enipCommand = 0x0065
+	unregisterSession enipCommand = 0x0066
+	sendRRData        enipCommand = 0x006f
+	sendUnitData      enipCommand = 0x0070
+
+	ErrUnknownENIPCommand = errors.New("Unknown Ethernet/IP Command")
+	ErrENIPDataTooSmall   = errors.New("ENIP packet data truncated")
+)
+
+type enipCommand uint16
+
+type ENIP struct {
+	BaseLayer
+	Command         enipCommand
+	Length          uint16
+	SessionHandle   uint32
+	Status          uint32
+	SenderContext   []byte
+	Options         uint32
+	CommandSpecific ENIPCommandSpecificData
```

Comment:
Here, I'd consider adding the enipCommand to the top-level struct, and putting CommandSpecific.Data in a wrapped Payload layer.